src/Npgsql/MultiHostConnectorPool.cs

```diff
@@ -0,0 +1,261 @@
+                {
+                    exceptions.Add(new NpgsqlException($"Unable to connect to {pool.Settings.Host}:{pool.Settings.Port}", ex));
+                }
+            }
+
+            return null;
+        }
+
+        async ValueTask<NpgsqlConnector?> TryRent(NpgsqlConnection conn, TimeSpan timeoutPerHost, bool async, TargetServerType preferedType,
+            Func<ConnectorPool, TargetServerType, bool> validHostFunc, IList<Exception> exceptions,
+            CancellationToken cancellationToken)
+        {
+            foreach (var pool in _pools)
+            {
+                if (!validHostFunc(pool, preferedType))
+                    continue;
+
+                NpgsqlConnector? connector = null;
+
+                try
+                {
+                    connector = await pool.Rent(conn, new NpgsqlTimeout(timeoutPerHost), async, cancellationToken);
+                    if (GetHostState(pool) == HostState.Unknown)
+                        await connector.UpdateServerType(new NpgsqlTimeout(timeoutPerHost), async, cancellationToken);
+
+                    if (!validHostFunc(pool, preferedType))
+                    {
+                        conn.Connector = null;
+                        connector.Connection = null;
+                        pool.Return(connector);
+                        continue;
+                    }
+
+                    conn.Connector = connector;
+                    connector.Connection = conn;
+                    return connector;
+                }
+                catch (Exception ex)
+                {
+                    if (connector is not null)
+                    {
+                        conn.Connector = null;
+                        connector.Connection = null;
+                        pool.Return(connector);
+                    }
+
+                    exceptions.Add(new NpgsqlException($"Unable to connect to {pool.Settings.Host}:{pool.Settings.Port}", ex));
+                }
+            }
+
+            return null;
+        }
+
+        internal override async ValueTask<NpgsqlConnector> Rent(NpgsqlConnection conn, NpgsqlTimeout timeout, bool async, CancellationToken cancellationToken)
+        {
+            var exceptions = new List<Exception>();
+
+            var timeoutPerHost = timeout.IsSet ? timeout.CheckAndGetTimeLeft() : TimeSpan.Zero;
+            var preferedType = conn.Settings.TargetServerType;
+
+            var idlePreferedConnector = await TryGetIdle(conn, timeoutPerHost, async, preferedType, IsValidHost, exceptions, cancellationToken);
+            if (idlePreferedConnector is not null)
+                return idlePreferedConnector;
+
+            var newPreferedConnector = await TryOpenNew(conn, timeoutPerHost, async, preferedType, IsValidHost, exceptions, cancellationToken);
+            if (newPreferedConnector is not null)
+                return newPreferedConnector;
+
+            if (preferedType == TargetServerType.Any)
```

Comment:
I think this means that if Any is specified, we might block on some exhausted pool, where if we didn't specify Any, we'd get an idle non-preferred connection below... Am I understanding right?

Reply:
Nope, it just attempts to rent a connection from any non-offline pool.

Reply:
Right, the point I was missing is that for Any, there's no sense in going into the fallback flow, because Any has no fallback.